File: 5.最长回文子串.js
/*
 * @lc app=leetcode.cn id=5 lang=javascript
 *
 * [5] 最长回文子串
 */

// @lc code=start
/**
 * @param {string} s
 * @return {string}
 */

// 103/103 cases passed (88 ms)
// Your runtime beats 88.06 % of javascript submissions
// Your memory usage beats 62.64 % of javascript submissions (37.1 MB)
function caluLongestPalindrome(str, left, right) {
  if (
    left < 0 ||
    right < 0 ||
    left >= str.length||
    right >= str.length
  ) {
    // throw exception
  }
  while (
    left >= 0 &&
    right < str.length &&
    str.charAt(left) == str.charAt(right)
  ) {
    left--;
    right++;
  }
  return [left + 1, right];
}

var longestPalindrome = function(s) {
  if (s.length == 0) return "";
  let candidate = s.charAt(0);
  let tmp = "";
  for (let i = 0; i < s.length; i++) {
    tmp = caluLongestPalindrome(s, i, i);
    candidate = candidate.length < tmp[1]-tmp[0] ? s.substring(tmp[0], tmp[1]) : candidate;
    if (i + 1 < s.length) {
      tmp = caluLongestPalindrome(s, i, i + 1);
      candidate = candidate.length < tmp[1]-tmp[0] ? s.substring(tmp[0], tmp[1]) : candidate;
    }
  }
  return candidate;
};

// 103/103 cases passed (104 ms)
// Your runtime beats 69.13 % of javascript submissions
// Your memory usage beats 33.25 % of javascript submissions (41.7 MB)
// var longestPalindrome = function(s) {
//     let last_index = [0];
//     let curr_index = [];
//     let candidate = s.length > 0 ? s.substring(0,1) : "";
//     for(let i = 1; i<s.length; i++){
//         curr_index[0]= i;
//         if(s.charAt(i) == s.charAt(i -1)){
//             curr_index[1]= i - 1;
//             candidate = candidate.length < 2 ? s.substring(i -1, i +1) : candidate;
//         }
//         for(let j = 0; j < last_index.length; j++){
//             if(last_index[j] > 0 && s.charAt(i) == s.charAt(last_index[j] -1)){
//                 curr_index.push(last_index[j] -1);
//                 if(i - last_index[j] + 2 > candidate.length){
//                     candidate = s.substring(last_index[j] -1, i +1);
//                 }
//             }
//         }
//         last_index = curr_index;
//         curr_index = [];
//     }
//     return candidate;
// };
// @lc code=end

console.log(longestPalindrome("cbbd"));
console.log("end");

File: 5.最长回文子串.java
/*
 * @lc app=leetcode.cn id=5 lang=java
 *
 * [5] 最长回文子串
 */

// @lc code=start
class Solution {
    // public String longestPalindrome(String s) {
    //     if(s.length()<=0){
    //         return "";
    //     }
    //     int pointX = 0;
    //     int pointY = 0;
    //     int[][] arr = new int[s.length()][s.length()];   
    //     for(int i = 0; i < s.length(); i++){
    //         for(int j = 0; j < s.length(); j++){
    //             boolean isSame = s.charAt(i) == s.charAt(s.length() -1 - j);
    //             // 存储二维数组
    //             if(!isSame){
    //                 arr[i][j] = 0;
    //             } else if(i == 0 || j == 0){
    //                 arr[i][j] = 1;
    //             } else {
    //                 arr[i][j] = arr[i - 1][j - 1] + 1;
    //             }
    //             // 存储最长的点
    //             if(arr[pointX][pointY] < arr[i][j] ){
    //                 pointX = i;
    //                 pointY = j;
    //             }
    //         } 
    //     }
    //     return s.substring(pointX - arr[pointX][pointY] + 1, pointX + 1);
    // }
    
    private String getLongestPalindrome(String base, int left, int right) {
        if (left < 0 || right < 0 || left >= base.length() || right >= base.length()) {
            // throw exception
        }
        while(left >= 0 && right < base.length() && base.charAt(left) == base.charAt(right)) {
            left --;
            right ++;
        }
        return base.substring(left + 1, right);
    }
    public String longestPalindrome(String s) {
        String candidate = "";
        for (int i = 0; i < s.length(); i++) {
            String tmp = getLongestPalindrome(s, i, i);
            System.out.println(tmp);
            if (tmp.length() > candidate.length()) {
                candidate = tmp;
            }
            if ((i + 1) < s.length()) {
                tmp = getLongestPalindrome(s, i, i + 1);
                System.out.println(tmp);
                if (tmp.length() > candidate.length()) {
                    candidate = tmp;
                }
            }
        }
        return candidate;
    }
    public String longestPalindrome(String s) {
        String candidate = "";
        int last_index[] = new int[s.length()];
        int curr_index[] = new int[s.length()];

        // int[] last_index;
        // int[] curr_index;
        // int last_index[] = new int[];
        // int curr_index[] = new int[s.length()];
        for(int i = 0; i<s.length(); i++){
            curr_index[0]= i;
            if(i == 0){
                // last_index = curr_index;
                candidate = s.substring(i, i+1);
            }else if(i > 0){
                for(int j = 0; j < last_index.length; j++){
                    // 0 没判断
                    // 记录 last_index 长度
                    if(s.charAt(i) == s.charAt(last_index[j] -1)){
                        // curr_index.push(last_index[j] -1);
                        curr_index[curr_index.length] = last_index[j] -1;
                        String tmp = s.substring(last_index[j] -1, i +1);
                        if(tmp.length() > candidate.length()){
                            candidate = s.substring(last_index[j] -1, i +1);
                        }
                    }
                }
            }
            last_index = curr_index; //
            curr_index = []; // 
        }
        return candidate;
    }
}
// @lc code=end

